### PR TITLE
cc: convert compiler wrappers to POSIX sh

### LINKED ISSUE #8703: Make Spack's compiler wrapper compatible with dash
Currently, Spack's compiler wrapper is hardcoded to use `/bin/bash`. However, if we changed this to `/bin/sh` and made things POSIX compliant, systems like Ubuntu that default to `dash` could see ~~vast~~ slight improvements to build times for large, complex packages. If we changed it to `/usr/bin/env sh`, Spack users could install `dash` on other systems and add it to their `PATH`, although I don't know how much `env` will slow down runtimes.

If no one else gets to this, I'll probably experiment with it eventually. I just wanted to make a dedicated issue before I forget.

Related to #4048. Probably depends on #7604. See https://unix.stackexchange.com/questions/148035 for performance comparisons between common shells.

### PR DESCRIPTION
**Pull request by tgamblin**
Fixes #8703.
Closes #25380.

Putting up an initial PR to start the process of converting `cc` to POSIX shell.

Remaining to do:
- [x] initial `shellcheck`-based fixes
- [ ] replace arrays with posix sh tricks
- [ ] `read`
- [ ] heredocs (is this needed?)
- [ ] additional fixes

### CODE REVIEW

**Comment by tgamblin** [comment]
@haampie: what is the issue with heredocs?  They're supported by Bourne shell so not sure why they're listed as needed.

**Comment by tgamblin** [comment]
Looks like the initial changes weren't entirely sound -- tests are failing.

**Review by christoph-conrads — COMMENTED** [review]
POSIX shells do not support arrays, cf. [StackOverflow: _Arrays in unix shell?_](https://stackoverflow.com/a/34183853) Error message on Alpine Linux with Ash:
```
/home/john/spack/lib/spack/env/gcc/gcc: line 26: syntax error: unexpected "("
```

File: lib/spack/env/cc [hunk 1]
```diff
@@ -78,10 +78,10 @@ IFS=' ' read -ra SPACK_LDFLAGS  <<< "$SPACK_LDFLAGS"
 IFS=' ' read -ra SPACK_LDLIBS   <<< "$SPACK_LDLIBS"
```
Comment (christoph-conrads) [inline lib/spack/env/cc:78]: * Here strings (`<<<`) are a POSIX feature
* `read -a` creates an array; POSIX shells do not feature arrays; dash complains with `dash: 1: read: Illegal option -a`

**Review by christoph-conrads — COMMENTED** [review]
POSIX shells do not support arrays, cf. [StackOverflow: _Arrays in unix shell?_](https://stackoverflow.com/a/34183853) Error message on Alpine Linux with Ash:
```
/home/john/spack/lib/spack/env/gcc/gcc: line 26: syntax error: unexpected "("
```

**Comment by christoph-conrads** [comment]
> Putting up an initial PR to start the process of converting `cc` to POSIX shell.

Why not Python?

* There must be a Python interpreter available; you cannot run Spack without it
* A lot of features:
  * assertions
  * data structures like arrays, maps, and so on
  * optional static typing
  * exceptions and/or proper return values
* Absence of error sources
  * confusion between quoted/unquoted parameter substitution, parameters with/without newlines
  * the interaction between `set -e` and using the exit status as a branching condition

**Comment by tgamblin** [comment]
> POSIX shells do not support arrays

Yep, this is about removing arrays and emulating them -- see http://www.etalabs.net/sh_tricks.html.

**Comment by tgamblin** [comment]
@christoph-conrads:
> Why not Python?

It'd be great if we could.  In fact, we started this way.  If you go back to 0f3b80cddbfd775c10f68423444aca792525f856 (in 2014), you'll see that we replaced the then-python version with `bash`.  This was done because `python` startup overhead is really high and was making builds extremely slow.  We frequently saw slowdowns of ~6x for builds, just from the wrappers.

**Comment by christoph-conrads** [comment]
> Yep, this is about removing arrays and emulating them -- see http://www.etalabs.net/sh_tricks.html.

I see.

> Unlike “enhanced” Bourne shells such as Bash, the POSIX shell does not have array types. However, with a bit of inefficiency, you can get array-like semantics in a pinch using pure POSIX sh. The trick is that you do have one (and only one) array — the positional parameters “$1”, “$2”, etc. — and you can swap things in and out of this array.

:grimacing:

Have you considered having the shell code generated?

**Comment by christoph-conrads** [comment]
> @christoph-conrads:
> 
> > Why not Python?
> 
> It'd be great if we could. In fact, we started this way. If you go back to [0f3b80c](https://github.com/spack/spack/commit/0f3b80cddbfd775c10f68423444aca792525f856) (in 2014), you'll see that we replaced the then-python version with `bash`. This was done because `python` startup overhead is really high and was making builds extremely slow. We frequently saw slowdowns of ~6x for builds, just from the wrappers.

Thank you for the link to the commit.

**Comment by haampie** [comment]
@christoph-conrads see https://github.com/spack/spack/pull/25380#discussion_r689940544 and below

**Comment by christoph-conrads** [comment]
> @christoph-conrads see [#25380 (comment)](https://github.com/spack/spack/pull/25380#discussion_r689940544) and below

[From the discussion](https://github.com/spack/spack/pull/25380#issuecomment-901049740):

Generating the compiler wrapper, benchmarks, everything. You've discussed all approaches that came to my mind.

**Comment by tgamblin** [comment]
closing this as @cosmicexplorer is working on #26048.

**Pull request closed without merging**